src/websocketmiddleware.ts

```diff
@@ -0,0 +1,84 @@
+import { Dispatch } from "react";
+import { AnyAction, Middleware } from "redux";
+import { API_ROOT } from "./api/util";
+import { WebSocketAction, WebSocketActionType } from "./store/actions/types";
+
+const SOCKET_ROOT = (() => {
+    const url = new URL(API_ROOT);
+    url.protocol = "ws:";
+    return url;
+})();
+
+const registeredSockets: Record<string, WebSocket | null> = {};
+
+function openWebSocket(dispatch: Dispatch<WebSocketAction>, action: WebSocketAction): WebSocket {
+    if(action.type !== WebSocketActionType.OPEN) {
+        throw Error(`openWebSocket expects a ${WebSocketActionType.OPEN} action, got ${action.type}`);
+    }
+
+    if(registeredSockets[action.name] != null) {
+        throw Error(`An open websocket with name ${action.name} already exists.`);
+    }
+
+    const socket = new WebSocket(`${SOCKET_ROOT}${action.endpoint}?token=${action.token}`);
+    socket.addEventListener('open', () => {
+        registeredSockets[action.name] = socket;
+    });
+
+    socket.addEventListener('message', (e: MessageEvent) => {
+        dispatch({
+            type: WebSocketActionType.RECEIVE,
+            name: action.name,
+            message: e.data,
+            event: e
+        });
+    });
+
+    socket.addEventListener('close', (e: CloseEvent) => {
+        registeredSockets[action.name] = null;
+        dispatch({
+            type: WebSocketActionType.CLOSED,
+            name: action.name,
+            event: e
+        });
+    });
+
+    return socket;
+}
+
+function sendWebSocketMessage(name: string, message: string) {
+    const socket = registeredSockets[name];
+    if(socket == null) {
+        throw Error("WebSocket was closed.");
+    }
+    socket.send(message);
+}
+
+function handleWebSocketAction(next: Dispatch<WebSocketAction>, action: WebSocketAction) {
+    switch(action.type) {
+        case WebSocketActionType.OPEN:
+            openWebSocket(next, action);
+            break;
+        case WebSocketActionType.SEND:
+            sendWebSocketMessage(action.name, action.message);
```

Comment:
Shouldn't you break this statement just for good practice? It will FALL THROUGH :open_mouth:

Reply:
I mean break the case